iree/compiler/Codegen/LLVMCPU/VectorContractCustomKernels.cpp

```diff
@@ -321,18 +319,137 @@ struct MMT_8x4x8_i8i8i32_Aarch64Dotprod_InlineAsm
+    auto rhs = contractionOp.rhs();
+    if (acc.getType().cast<VectorType>().getElementType() != I32Type) {
+      return failure();
+    }
+
+    Value inLhs = getExtSIInput(I8Type, I32Type, lhs);
+    Value inRhs = getExtSIInput(I8Type, I32Type, rhs);
+
+    if (!inLhs || !inRhs) return failure();
+
+    auto loc = contractionOp.getLoc();
+
+    auto int32x4VType = VectorType::get({4}, I32Type);
+
+    std::array<Value, 16> accChunks;
+    for (int row = 0; row < 8; ++row) {
+      auto accRow =
+          rewriter.create<vector::ExtractOp>(loc, acc, ArrayRef<int64_t>{row});
+      for (int chunk = 0; chunk < 2; ++chunk) {
+        auto accChunk = rewriter.create<vector::ExtractStridedSliceOp>(
+            loc, accRow, ArrayRef<int64_t>{chunk * 4}, ArrayRef<int64_t>{4},
+            ArrayRef<int64_t>{1});
+        assert(accChunk.getType() == int32x4VType);
+        accChunks[row * 2 + chunk] = accChunk;
+      }
+    }
+
+    auto int8x4x4VType = VectorType::get({4, 4}, rewriter.getIntegerType(8));
+    auto extract4x4 = [&](Value in, int rowOffset, int colOffset) {
+      auto chunk = rewriter.create<vector::ExtractStridedSliceOp>(
+          loc, in, ArrayRef<int64_t>{rowOffset, colOffset},
+          ArrayRef<int64_t>{4, 4}, ArrayRef<int64_t>{1, 1});
+      assert(chunk.getType() == int8x4x4VType);
+      return chunk;
+    };
+
+    auto lhs0 = extract4x4(inLhs, 0, 0);
+    auto lhs1 = extract4x4(inLhs, 4, 0);
+    auto rhs0 = extract4x4(inRhs, 0, 0);
+    auto rhs1 = extract4x4(inRhs, 4, 0);
+
+    auto int8Zero4x4 = rewriter.create<arith::ConstantOp>(
+        loc, rewriter.getZeroAttr(int8x4x4VType));
+    auto sdot = [&](Value acc, Value a, Value b, int64_t lane) -> Value {
+      auto bReplicatedLane = rewriter.create<vector::ShuffleOp>(
+          loc, b, int8Zero4x4, ArrayRef<int64_t>{lane, lane, lane, lane});
+
+      return rewriter.create<arm_neon::Sdot2dOp>(loc, int32x4VType, acc, a,
+                                                 bReplicatedLane);
+    };
+
+    std::array<Value, 16> dstChunks;
+    // Note lhs and rhs are deliberately swapped here
+    dstChunks[0] = sdot(accChunks[0], rhs0, lhs0, 0);
+    dstChunks[1] = sdot(accChunks[1], rhs1, lhs0, 0);
+    dstChunks[2] = sdot(accChunks[2], rhs0, lhs0, 1);
+    dstChunks[3] = sdot(accChunks[3], rhs1, lhs0, 1);
+    dstChunks[4] = sdot(accChunks[4], rhs0, lhs0, 2);
+    dstChunks[5] = sdot(accChunks[5], rhs1, lhs0, 2);
+    dstChunks[6] = sdot(accChunks[6], rhs0, lhs0, 3);
+    dstChunks[7] = sdot(accChunks[7], rhs1, lhs0, 3);
+    dstChunks[8] = sdot(accChunks[8], rhs0, lhs1, 0);
+    dstChunks[9] = sdot(accChunks[9], rhs1, lhs1, 0);
+    dstChunks[10] = sdot(accChunks[10], rhs0, lhs1, 1);
+    dstChunks[11] = sdot(accChunks[11], rhs1, lhs1, 1);
+    dstChunks[12] = sdot(accChunks[12], rhs0, lhs1, 2);
+    dstChunks[13] = sdot(accChunks[13], rhs1, lhs1, 2);
+    dstChunks[14] = sdot(accChunks[14], rhs0, lhs1, 3);
+    dstChunks[15] = sdot(accChunks[15], rhs1, lhs1, 3);
```

Comment:
it would be possible to roll these 16 lines into a simple for loop (if you put `lhs0` and `lhs1` into a `lhs[2]` array). but i'm also cool with leaving it unrolled if you like how it makes this c++ code look more like the resulting kernel. note though, we might move toward some genericity, as there will be some variations around this current 8x8 shape --- in both intrinsics and asm kernels. Variation could be motivated by either narrow matmuls (e.g. need a 8x1 kernel for matrix-times-vector cases) or to allow experimenting with getting the last percent of peak performance on some hardware target (so we might actually play with a 12x8 instead of 8x8 shape).

Reply:
Sure, done